src/linalg.jl

```diff
@@ -629,6 +629,61 @@ function dot(A::AbstractSparseMatrixCSC, B::Union{DenseMatrixUnion,WrapperMatrix
     return conj(dot(B, A))
 end
 
+function dot(x::AbstractSparseVector, Q::Diagonal, y::AbstractVector)
+    d = Q.diag
+    if length(x) != length(y) || length(y) != length(d)
+        throw(
+            DimensionMismatch("Vectors and matrix have different dimensions, x has a length $(length(x)), y has a length $(length(y)), Q has side dimension $(size(Q, 1))")
+        )
+    end
```

Comment:
```suggestion
    length(x) == length(y)  == length(d) ||
        throw(
            DimensionMismatch("Vectors and matrix have different dimensions, x has a length $(length(x)), y has a length $(length(y)), Q has side dimension $(size(Q, 1))")
        )
```
?

Same below, since you're not reusing `n = length(x)` there.

Reply:
just changed